examples/attention/Globals.cc

```diff
@@ -0,0 +1,163 @@
+#include "Globals.h"
+
+using namespace opencog;
+using namespace opencog::ECANExperiment;
+
+std::vector<std::vector<std::string>> opencog::ECANExperiment::swords;
+std::vector<std::string> opencog::ECANExperiment::words;
+
+int opencog::ECANExperiment::current_group = 0;
+
+int opencog::ECANExperiment::special_word_occurence_period = 1;
+
+bool opencog::ECANExperiment::are_similar(const Handle& h1, const Handle& h2, bool strict_type_match)
+{
+    if (h1 == h2)
+        return true;
+
+    if (NodeCast(h1) and NodeCast(h2))
+        return !strict_type_match or h1->getType() == h2->getType();
+
+    LinkPtr lh1(LinkCast(h1));
+    LinkPtr lh2(LinkCast(h2));
+
+    if (lh1 and lh2) {
+        if (strict_type_match and (lh1->getType() != lh2->getType()))
+            return false;
+
+        HandleSeq hseqh1 = lh1->getOutgoingSet();
+        HandleSeq hseqh2 = lh2->getOutgoingSet();
+
+        if (hseqh1.size() != hseqh2.size())
+            return false;
+
+        // Unordered links should be treated in a special way
+        if (classserver().isA(lh1->getType(), UNORDERED_LINK) or classserver().isA(
+                lh2->getType(), UNORDERED_LINK)) {
+
+            for (const auto& h1 : hseqh1) {
+                for (auto it = hseqh2.begin(); it != hseqh2.end(); ++it) {
+                    if (are_similar(h1, h2, strict_type_match)) {
+                        hseqh2.erase(it);
+                        break;
+                    }
+                }
+            }
+
+            // Empty means all has been mapped. Success.
+            return hseqh2.empty() or false;
+        }
+
+        for (HandleSeq::size_type i = 0; i < hseqh1.size(); i++) {
+            if (not are_similar(hseqh1[i], hseqh2[i], strict_type_match))
+                return false;
+        }
+
+        return true;
+    }
+
+    return false;
+}
+
+bool opencog::ECANExperiment::exists_in(const Handle& hlink, const Handle& h)
+{
+    if (hlink == h) {
+        return true;
+    } else {
+        LinkPtr lp(LinkCast(hlink));
+        if (nullptr == lp)
+            return false;
+
+        auto outg = lp->getOutgoingSet();
+        if (std::find(outg.begin(), outg.end(), h) != outg.end())
+            return true;
+        else {
+            for (const Handle& hi : outg) {
+                if (LinkCast(hi) and exists_in(hi, h))
+                    return true;
+            }
+        }
+        return false;
+    }
+}
+
+bool opencog::ECANExperiment::is_friendship_reln(const Handle& h)
+{
+    AtomSpace as;
+    as.add_atom(h);
+    Handle friends_predicate = as.add_node(PREDICATE_NODE, "friends");
+    Handle var_1 = as.add_node(CONCEPT_NODE, "$0343O45FFEWW");
+    Handle var_2 = as.add_node(CONCEPT_NODE, "$0343045FFEWY");
+    Handle friend_list = as.add_link(LIST_LINK, HandleSeq { var_1, var_2 });
+    Handle eval_link = as.add_link(EVALUATION_LINK, { friends_predicate,
+                                                      friend_list });
+    as.add_atom(eval_link);
```

Comment:
Huh ... this is exactly why we have python and scheme bindings (and haskell!) -- its because adding atoms to the atomspace in C++ is verbose and painful, and any changes require a recompile.